tests/test-mongoc-command-monitoring.c

```diff
@@ -1745,6 +1745,144 @@ test_killcursors_deprecated (void)
+      /* the test is ending */
+      return;
+   }
+
+   test = (cmd_test_t *) mongoc_apm_command_started_get_context (event);
+   test->started_calls++;
+   bson_destroy (&test->cmd);
+   bson_strncpy (test->db,
+                 mongoc_apm_command_started_get_database_name (event),
+                 sizeof (test->db));
+   bson_copy_to (mongoc_apm_command_started_get_command (event), &test->cmd);
+   bson_strncpy (test->cmd_name,
+                 mongoc_apm_command_started_get_command_name (event),
+                 sizeof (test->cmd_name));
+}
+
+
+static void
+command_failed_reply_command_failed_cb (const mongoc_apm_command_failed_t *event)
+{
+   cmd_test_t *test;
+
+   test = (cmd_test_t *) mongoc_apm_command_failed_get_context (event);
+   test->failed_calls++;
+   bson_destroy (&test->cmd);
+
+   ASSERT_CMPSTR (test->cmd_name,
+                  mongoc_apm_command_failed_get_command_name (event));
+
+   bson_copy_to (mongoc_apm_command_failed_get_reply (event), &test->cmd);
+}
+
+
+static void
+test_command_failed_reply ()
+{
+   mock_server_t *server;
+   mongoc_client_t *client;
+   mongoc_apm_callbacks_t *callbacks;
+   mongoc_collection_t *collection;
+   cmd_test_t test;
+   mongoc_cursor_t *cursor;
+   const bson_t *doc;
+   future_t *future;
+   request_t *request;
+
+   /* test that the command_failed_event's reply is the same as a mocked reply*/
+   cmd_test_init (&test);
+
+   server = mock_server_with_autoismaster (4);
+   mock_server_run (server);
+
+   callbacks = mongoc_apm_callbacks_new ();
+   mongoc_apm_set_command_started_cb(callbacks, command_failed_reply_command_started_cb);
+   mongoc_apm_set_command_failed_cb (callbacks, command_failed_reply_command_failed_cb);
+
+   client = mongoc_client_new_from_uri (mock_server_get_uri (server));
+   ASSERT (mongoc_client_set_apm_callbacks (client, callbacks, (void *) &test));
+
+   collection = mongoc_client_get_collection (client, "db", "collection");
+   cursor = mongoc_collection_find (
+      collection, MONGOC_QUERY_NONE, 0, 0, 1, tmp_bson ("{}"), NULL, NULL);
+
+   future = future_cursor_next (cursor, &doc);
+   request = mock_server_receives_request (server);
+   mock_server_replies_simple (request,
+                               "{'ok': 0, 'code': 42, 'errmsg': 'bad!'}");
+
+   ASSERT (!future_get_bool (future));
+   future_destroy (future);
+   request_destroy (request);
+
+   ASSERT_MATCH (&test.cmd, "{'ok': 0, 'code': 42, 'errmsg': 'bad!'}");
+   ASSERT_CMPINT (test.failed_calls, ==, 1);
+
+   mock_server_destroy (server);
+
+   /* client logs warning because it can't send killCursors or endSessions */
+   capture_logs (true);
+   mongoc_cursor_destroy (cursor);
+   mongoc_collection_destroy (collection);
+
+   mongoc_client_destroy (client);
+
+   cmd_test_cleanup (&test);
+   mongoc_apm_callbacks_destroy (callbacks);
+
+   /* test that the command_failed_event's reply is empty if there is a network error (i.e. the server hangs up) */
+   cmd_test_init (&test);
```

Comment:
let's make this two test functions to test the two scenarios - starting here it seems you reinit all the variables anyway